### ISSUE TITLE
Finding a solution for etcd

### ISSUE THREAD

**Issue by jamiehannaford** [kubernetes/kubeadm#277]
A few weeks a go some folks on Slack brought up the idea of defining requirements for highly available etcd on kubeadm-provisioned clusters. The idea was to define requirements before the implementation of any solution. This discussion was continued in the sig-cluster-lifecycle meeting on [May 16th 2017](https://docs.google.com/document/d/1deJYPIF4LmhGjDVaqrswErIrV7mtwJgovtLnPCDxP7U/edit#heading=h.50f520r1aizs), where we came up with some initial critieria:

1. High availability
  a. Recovers from member failure
  b. Recovers from quorum loss
  c. Recovery from full cluster failure i.e. power-off
  d. Recovers from partial / failed / interrupted upgrades
2. Handles discovery of etcd peers
3. Secure by default
   a. TLS encryption
   b. Certificate rotation
4. Support multiple form factors
   a. Non-self hosted
   b. Self-hosted (optional)
5. Ability to restore from a backup (possibly not backup itself)
6. Upgrades
   a. Rolling upgrades
   b. Downgrades ([but tricky because of etcd](https://github.com/coreos/etcd/issues/7308))
7. Resize/scale cluster from 1 -> 3 -> 5 members
8. Ease of installation/teardown

Are there any I've missed?

The next stage is proposing solutions that meet the above criteria and can be verified in a fork.

cc/ @timothysc @justinsb @philips @xiang90 @aaronlevy

**Comment by jamiehannaford** [comment]
My proposed solution is the [etcd-operator](https://github.com/coreos/etcd-operator), since it provides nearly all criteria out of the box:

- High availability
  - [x] Recovers from member failure (see [here](https://github.com/coreos/etcd-operator#member-recovery))
  - [x] Recovers from quorum loss (see [here](https://github.com/coreos/etcd-operator#disaster-recovery))
  - [x] Recovery from full cluster failure i.e. power-off (I think this can be accomplished using snapshotting and the operator's built in [restart tolerance](https://github.com/coreos/etcd-operator#etcd-operator-recovery))
  - [ ] Recovers from partial / failed / interrupted upgrades
- [x] Handles discovery of etcd peers
- Secure by default
  - [x] TLS encryption (see [here](https://github.com/coreos/etcd-operator/blob/master/pkg/spec/cluster_tls.go#L20-L24))
  - [ ] Certificate rotation (no, but my personal preference is to use an external tool, like a cert-rotator operator to handle TLS cert rotation across the cluster)
- [x] Support multiple form factors
  - [x] Non-self hosted 
  - [x] Self-hosted 
- [x] Ability to restore from a backup (see [here](https://github.com/coreos/etcd-operator#disaster-recovery))
- [x] Upgrades
  - [x] Rolling upgrades (see [here](https://github.com/coreos/etcd-operator#upgrade-an-etcd-cluster))
  - [x] Downgrades (but tricky because of etcd)
- [x] Resize/scale cluster from 1 -> 3 -> 5 members (see [here](https://github.com/coreos/etcd-operator#resize-an-etcd-cluster))
- [x] Ease of installation/teardown 

I've already integrated kubeadm and etcd-operator successfully in [this PR](https://github.com/kubernetes/kubernetes/pull/45665), and [here](https://github.com/jamiehannaford/kubernetes/tree/etcd-operator/cmd/kubeadm) is the fork. 

I think it's probably worthwhile to come up with a more granular disaster recovery requirement list, and also think about to what degree an etcd solution should cover all the bases. We already have several issues where this is tracked, so we should take into account all the suggested solutions there too:

- https://github.com/kubernetes-incubator/bootkube/issues/432
- https://github.com/kubernetes-incubator/bootkube/issues/112
- https://github.com/coreos/tectonic-installer/issues/257

**Comment by xiang90** [comment]
> Recovers from partial / failed / interrupted upgrades

see https://github.com/kubernetes-incubator/bootkube/pull/528

**Comment by timothysc** [comment]
@jamiehannaford if there is a branch you would like reviewed I'd be happy to go through it now.

**Comment by jamiehannaford** [comment]
@xiang90 Awesome. So if a failed upgrade occurs, the user can manual restore from a backup file. Is there a way that etcd can automatically check specific locations (like local FS, S3) for backups without the user needing to specify one manually? 

For example, assume that the etcd-operator has been backing stuff up to a S3 container. When it initialises it checks the same bucket and boots from there (this assume the user hasn't changed any backup options).

**Comment by jamiehannaford** [comment]
@timothysc Thanks! The only branch I have is the one I submitted in [my PR](https://github.com/kubernetes/kubernetes/pull/45665) (on hold until we agree on a next step forward). I think you've already gone through this though. Unless you meant something else?

**Comment by luxas** [comment]
> There are a bunch of comments on that PR which I can start to address as a next step forward. I think I'll also add TLS secrets to the PR too. Should I go ahead and do that?

@jamiehannaford Feel free to. I'm gonna try to look at the TLS Secrets PR this week so it might yet change (@andrewrynhard), but I don't expect it to be part of v1.7 to give us a little more time to think about it until v1.8

**Comment by anguslees** [comment]
> For example, assume that the etcd-operator has been backing stuff up to a S3 container. When it initialises it checks the same bucket and boots from there

I'd just like to highlight that doing something like this *automatically* is a terrible idea and will give you multiple sources of truth if etcd is internally partitioned.

I suspect recovery will need to be manually triggered, because by definition it is required when the etcd cluster is incapable of making robust automatic decisions.

**Comment by xiang90** [comment]
> I'd just like to highlight that doing something like this automatically is a terrible idea and will give you multiple sources of truth if etcd is internally partitioned.

totally agree. we designed this to be a manual work, at least at etcd operator side.

**Comment by luxas** [comment]
Moving milestone to v1.9. In v1.8, we're gonna stick with a local etcd instance listening on localhost.

**Comment by luxas** [comment]
Moving milestone for this to v1.10 as we depend on changes being made to the operator before we can use it and the code freeze is coming up.

**Comment by timothysc** [comment]
Given all the history here and recent feedback, we need to go with the non-operator option.

**Comment by timothysc** [comment]
So I'm going to close this issue and open a new one to outline the doc on using the existing commands to lay down etcd, we will likely have to wait until we have some of the other phases work done as well. 

/cc @chuckha @fabriziopandini @stealthybox
